File: packages/python/src/epaper_dithering/palettes.py
"""Color palettes and schemes for e-paper displays."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum


@dataclass(frozen=True)
class ColorPalette:
    """Color palette for an e-paper display."""

    colors: dict[str, tuple[int, int, int]]  # name -> RGB tuple
    accent: str  # Primary accent color name


class ColorScheme(Enum):
    """Common e-ink display color schemes.

    Each scheme stores a color palette optimized for e-paper displays.
    The integer values match OpenDisplay firmware conventions for compatibility.

    Usage:
        scheme = ColorScheme.BWR
        scheme.value           # 1 (firmware value)
        scheme.name            # "BWR"
        scheme.palette.colors  # {'black': (0,0,0), 'white': (255,255,255), 'red': (255,0,0)}
        scheme.accent_color    # "red"
    """

    MONO = (
        0,
        ColorPalette(
            colors={
                "black": (0, 0, 0),
                "white": (255, 255, 255),
            },
            accent="black",
        ),
    )

    BWR = (
        1,
        ColorPalette(
            colors={
                "black": (0, 0, 0),
                "white": (255, 255, 255),
                "red": (255, 0, 0),
            },
            accent="red",
        ),
    )

    BWY = (
        2,
        ColorPalette(
            colors={
                "black": (0, 0, 0),
                "white": (255, 255, 255),
                "yellow": (255, 255, 0),
            },
            accent="yellow",
        ),
    )

    BWRY = (
        3,
        ColorPalette(
            colors={
                "black": (0, 0, 0),
                "white": (255, 255, 255),
                "yellow": (255, 255, 0),
                "red": (255, 0, 0),
            },
            accent="red",
        ),
    )

    BWGBRY = (
        4,
        ColorPalette(
            colors={
                "black": (0, 0, 0),
                "white": (255, 255, 255),
                "yellow": (255, 255, 0),
                "red": (255, 0, 0),
                "blue": (0, 0, 255),
                "green": (0, 255, 0),
            },
            accent="red",
        ),
    )

    GRAYSCALE_4 = (
        5,
        ColorPalette(
            colors={
                "black": (0, 0, 0),
                "gray1": (85, 85, 85),
                "gray2": (170, 170, 170),
                "white": (255, 255, 255),
            },
            accent="black",
        ),
    )

    # NOTE: Values 6 and 7 are placeholders pending firmware assignment.
    GRAYSCALE_8 = (
        6,
        ColorPalette(
            colors={
                "black": (0, 0, 0),
                "gray1": (36, 36, 36),
                "gray2": (73, 73, 73),
                "gray3": (109, 109, 109),
                "gray4": (146, 146, 146),
                "gray5": (182, 182, 182),
                "gray6": (219, 219, 219),
                "white": (255, 255, 255),
            },
            accent="black",
        ),
    )

    GRAYSCALE_16 = (
        7,
        ColorPalette(
            colors={
                "black": (0, 0, 0),
                "gray1": (17, 17, 17),
                "gray2": (34, 34, 34),
                "gray3": (51, 51, 51),
                "gray4": (68, 68, 68),
                "gray5": (85, 85, 85),
                "gray6": (102, 102, 102),
                "gray7": (119, 119, 119),
                "gray8": (136, 136, 136),
                "gray9": (153, 153, 153),
                "gray10": (170, 170, 170),
                "gray11": (187, 187, 187),
                "gray12": (204, 204, 204),
                "gray13": (221, 221, 221),
                "gray14": (238, 238, 238),
                "white": (255, 255, 255),
            },
            accent="black",
        ),
    )

    def __init__(self, value: int, palette: ColorPalette):
        self._value_ = value  # type: ignore[assignment]
        self.palette = palette

    @property
    def accent_color(self) -> str:
        """Get accent color name for this scheme."""
        return self.palette.accent

    @property
    def color_count(self) -> int:
        """Get number of colors in palette."""
        return len(self.palette.colors)

    @classmethod
    def from_value(cls, value: int) -> ColorScheme:
        """Get ColorScheme from firmware int value.

        Args:
            value: Firmware color scheme value (0-7)

        Returns:
            Matching ColorScheme

        Raises:
            ValueError: If value is invalid
        """
        for scheme in cls:
            if scheme.value == value:  # type: ignore[comparison-overlap]
                return scheme
        raise ValueError(f"Invalid color scheme value: {value}")


# ============================================================================
# Measured Palettes for Specific E-Paper Displays
# ============================================================================
#
# These constants are derived from the Rust core at import time — RGB values
# are defined once in packages/rust/core/src/measured_palettes.rs.
#
# To add a new display palette:
#   1. Add a new `pub static` palette + CATALOG entry in measured_palettes.rs
#   2. Add the constant name here and export it in __init__.py
#   No RGB values needed in Python.
#
# ============================================================================


def _load_measured_palettes() -> dict[str, "ColorPalette"]:
    from . import _rs  # noqa: PLC0415  (local import avoids circular dependency at module level)

    result: dict[str, ColorPalette] = {}
    for name, rgb_bytes, color_names, accent_idx in _rs.measured_palettes():
        colors = {
            color_names[i]: (rgb_bytes[i * 3], rgb_bytes[i * 3 + 1], rgb_bytes[i * 3 + 2])
            for i in range(len(color_names))
        }
        result[name] = ColorPalette(colors=colors, accent=color_names[accent_idx])
    return result


_MEASURED = _load_measured_palettes()

SPECTRA_7_3_6COLOR: ColorPalette = _MEASURED["SPECTRA_7_3_6COLOR"]
SPECTRA_7_3_6COLOR_V2: ColorPalette = _MEASURED["SPECTRA_7_3_6COLOR_V2"]
MONO_4_26: ColorPalette = _MEASURED["MONO_4_26"]
BWRY_4_2: ColorPalette = _MEASURED["BWRY_4_2"]
BWRY_3_97: ColorPalette = _MEASURED["BWRY_3_97"]
SOLUM_BWR: ColorPalette = _MEASURED["SOLUM_BWR"]
HANSHOW_BWR: ColorPalette = _MEASURED["HANSHOW_BWR"]
HANSHOW_BWY: ColorPalette = _MEASURED["HANSHOW_BWY"]

File: packages/javascript/src/palettes.ts
import type { ColorPalette } from './types';

/**
 * E-paper display color schemes
 * Values match firmware conventions (0-7)
 */
export enum ColorScheme {
  MONO        = 0,
  BWR         = 1,
  BWY         = 2,
  BWRY        = 3,
  BWGBRY      = 4,
  GRAYSCALE_4 = 5,
  GRAYSCALE_16 = 6,
  /** Reserved: 8-level grayscale, pending firmware value assignment. */
  GRAYSCALE_8 = 7,
}

const PALETTES: Record<ColorScheme, ColorPalette> = {
  [ColorScheme.MONO]: {
    colors: {
      black: { r: 0, g: 0, b: 0 },
      white: { r: 255, g: 255, b: 255 },
    },
    accent: 'black',
  },
  [ColorScheme.BWR]: {
    colors: {
      black: { r: 0, g: 0, b: 0 },
      white: { r: 255, g: 255, b: 255 },
      red: { r: 255, g: 0, b: 0 },
    },
    accent: 'red',
  },
  [ColorScheme.BWY]: {
    colors: {
      black: { r: 0, g: 0, b: 0 },
      white: { r: 255, g: 255, b: 255 },
      yellow: { r: 255, g: 255, b: 0 },
    },
    accent: 'yellow',
  },
  [ColorScheme.BWRY]: {
    colors: {
      black: { r: 0, g: 0, b: 0 },
      white: { r: 255, g: 255, b: 255 },
      yellow: { r: 255, g: 255, b: 0 },
      red: { r: 255, g: 0, b: 0 },
    },
    accent: 'red',
  },
  [ColorScheme.BWGBRY]: {
    colors: {
      black: { r: 0, g: 0, b: 0 },
      white: { r: 255, g: 255, b: 255 },
      yellow: { r: 255, g: 255, b: 0 },
      red: { r: 255, g: 0, b: 0 },
      blue: { r: 0, g: 0, b: 255 },
      green: { r: 0, g: 255, b: 0 },
    },
    accent: 'red',
  },
  [ColorScheme.GRAYSCALE_4]: {
    colors: {
      black: { r: 0, g: 0, b: 0 },
      gray1: { r: 85, g: 85, b: 85 },
      gray2: { r: 170, g: 170, b: 170 },
      white: { r: 255, g: 255, b: 255 },
    },
    accent: 'black',
  },
  [ColorScheme.GRAYSCALE_8]: {
    colors: {
      black: { r: 0,   g: 0,   b: 0   },
      gray1: { r: 36,  g: 36,  b: 36  },
      gray2: { r: 73,  g: 73,  b: 73  },
      gray3: { r: 109, g: 109, b: 109 },
      gray4: { r: 146, g: 146, b: 146 },
      gray5: { r: 182, g: 182, b: 182 },
      gray6: { r: 219, g: 219, b: 219 },
      white: { r: 255, g: 255, b: 255 },
    },
    accent: 'black',
  },
  [ColorScheme.GRAYSCALE_16]: {
    colors: {
      black:  { r: 0,   g: 0,   b: 0   },
      gray1:  { r: 17,  g: 17,  b: 17  },
      gray2:  { r: 34,  g: 34,  b: 34  },
      gray3:  { r: 51,  g: 51,  b: 51  },
      gray4:  { r: 68,  g: 68,  b: 68  },
      gray5:  { r: 85,  g: 85,  b: 85  },
      gray6:  { r: 102, g: 102, b: 102 },
      gray7:  { r: 119, g: 119, b: 119 },
      gray8:  { r: 136, g: 136, b: 136 },
      gray9:  { r: 153, g: 153, b: 153 },
      gray10: { r: 170, g: 170, b: 170 },
      gray11: { r: 187, g: 187, b: 187 },
      gray12: { r: 204, g: 204, b: 204 },
      gray13: { r: 221, g: 221, b: 221 },
      gray14: { r: 238, g: 238, b: 238 },
      white:  { r: 255, g: 255, b: 255 },
    },
    accent: 'black',
  },
};

/** Get color palette for a color scheme */
export function getPalette(scheme: ColorScheme): ColorPalette {
  return PALETTES[scheme];
}

/** Get number of colors in a color scheme */
export function getColorCount(scheme: ColorScheme): number {
  return Object.keys(PALETTES[scheme].colors).length;
}

/** Create ColorScheme from firmware integer value */
export function fromValue(value: number): ColorScheme {
  if (value < 0 || value > 7) {
    throw new Error(`Invalid color scheme value: ${value}`);
  }
  return value as ColorScheme;
}

// =============================================================================
// Measured Palettes for Specific E-Paper Displays
// =============================================================================
//
// These constants provide measured RGB values from real e-paper displays.
// Pass them directly to ditherImage() instead of a ColorScheme enum.
//
// Usage:
//   import { ditherImage, SPECTRA_7_3_6COLOR } from '@opendisplay/epaper-dithering';
//   const result = ditherImage(imageBuffer, SPECTRA_7_3_6COLOR);
//
// NOTE: RGB values are defined in packages/rust/core/src/measured_palettes.rs
// (single source of truth). The Python package derives its constants from Rust
// via FFI at import time. TypeScript cannot do the same (WASM init order), so
// values here must be kept in sync manually with the Rust source.
// The WASM `measured_palettes()` function is exposed for future tooling.
//
// TO ADD A NEW DISPLAY: update measured_palettes.rs + add the constant below.
// =============================================================================

// 7.3" Spectra™ 6-color (BWGBRY scheme)
// Measured: 2026-02-03, iPhone 15 Pro Max RAW + Hue Play bars @ 6500K
// Paper reference RGB(215,217,218); normalization: value × (255/paper_channel)
export const SPECTRA_7_3_6COLOR: ColorPalette = {
  scheme: ColorScheme.BWGBRY,
  colors: {
    black:  { r: 26,  g: 13,  b: 35  },
    white:  { r: 185, g: 202, b: 205 },
    yellow: { r: 202, g: 184, b: 0   },
    red:    { r: 121, g: 9,   b: 0   },
    blue:   { r: 0,   g: 69,  b: 139 },
    green:  { r: 40,  g: 82,  b: 57  },
  },
  accent: 'red',
};

// 4.26" Monochrome (MONO scheme)
export const MONO_4_26: ColorPalette = {
  scheme: ColorScheme.MONO,
  colors: {
    black: { r: 5,   g: 5,   b: 5   },
    white: { r: 220, g: 220, b: 220 },
  },
  accent: 'black',
};

// 4.2" BWRY (BWRY scheme)
export const BWRY_4_2: ColorPalette = {
  scheme: ColorScheme.BWRY,
  colors: {
    black:  { r: 5,   g: 5,   b: 5   },
    white:  { r: 200, g: 200, b: 200 },
    yellow: { r: 200, g: 180, b: 0   },
    red:    { r: 120, g: 15,  b: 5   },
  },
  accent: 'red',
};

// Solum BWR (harvested display, BWR scheme)
export const SOLUM_BWR: ColorPalette = {
  scheme: ColorScheme.BWR,
  colors: {
    black: { r: 5,   g: 5,   b: 5   },
    white: { r: 200, g: 200, b: 200 },
    red:   { r: 120, g: 15,  b: 5   },
  },
  accent: 'red',
};

// Hanshow BWR (harvested display, BWR scheme)
export const HANSHOW_BWR: ColorPalette = {
  scheme: ColorScheme.BWR,
  colors: {
    black: { r: 5,   g: 5,   b: 5   },
    white: { r: 200, g: 200, b: 200 },
    red:   { r: 120, g: 15,  b: 5   },
  },
  accent: 'red',
};

// Hanshow BWY (harvested display, BWY scheme)
export const HANSHOW_BWY: ColorPalette = {
  scheme: ColorScheme.BWY,
  colors: {
    black:  { r: 5,   g: 5,   b: 5   },
    white:  { r: 200, g: 200, b: 200 },
    yellow: { r: 200, g: 180, b: 0   },
  },
  accent: 'yellow',
};

// 3.97" BWRY — EP397YR_800x480 (BWRY scheme)
// 7.3" Spectra™ 6-color (BWGBRY scheme) — v2 measurement
// Measured: 2026-03-15, iPhone 15 Pro Max RAW + Affinity (v3), A4 paper white reference
// Method: DNG with linear tone curve, WB from A4 paper, uniform ×2.4 scale
export const SPECTRA_7_3_6COLOR_V2: ColorPalette = {
  scheme: ColorScheme.BWGBRY,
  colors: {
    black:  { r: 31,  g: 24,  b: 41  },
    white:  { r: 168, g: 180, b: 182 },
    yellow: { r: 180, g: 173, b: 0   },
    red:    { r: 113, g: 24,  b: 19  },
    blue:   { r: 36,  g: 70,  b: 139 },
    green:  { r: 50,  g: 84,  b: 60  },
  },
  accent: 'red',
};

// Measured: 2026-03-06, iPhone RAW
// Paper reference RGB(205,205,205); normalization: value × (255/205)
export const BWRY_3_97: ColorPalette = {
  scheme: ColorScheme.BWRY,
  colors: {
    black:  { r: 10,  g: 7,   b: 14  },
    white:  { r: 173, g: 178, b: 174 },
    yellow: { r: 172, g: 128, b: 0   },
    red:    { r: 85,  g: 24,  b: 14  },
  },
  accent: 'red',
};
